databooks/cli.py

```diff
@@ -25,17 +27,54 @@
 app = Typer()
 
 
-def version_callback(value: bool) -> None:
+def _version_callback(show_version: bool) -> None:
     """Return application version."""
-    if value:
+    if show_version:
         echo("databooks version: " + _DISTRIBUTION_METADATA["Version"])
         raise Exit()
 
 
+def _help_callback(ctx: Context, show_help: Optional[bool]) -> None:
+    """Reimplement `help` command to execute eagerly."""
+    if show_help:
+        echo(ctx.command.get_help(ctx))
+        raise Exit()
+
+
+def _config_callback(ctx: Context, config_path: Optional[Path]) -> Optional[Path]:
+    """Get config file and inject values into context to override default args."""
+    target_paths = expand_paths(
+        paths=[Path(p) for p in ctx.params.get("paths", ())], rglob="*"
+    )
+    config_path = (
+        get_config(
+            target_paths=target_paths,
+            config_filename=TOML_CONFIG_FILE,
+        )
+        if config_path is None and target_paths
+        else config_path
+    )
+    logger.debug(f"Loading config file from: {config_path}")
+
+    ctx.default_map = ctx.default_map or {}  # initialize defaults
+
+    if config_path is not None:  # config may not be specified
+        with config_path.open("r") as f:
+            conf = (
+                tomli.load(f)
+                .get("tool", {})
+                .get("databooks", {})
+                .get(ctx.command.name, {})
+            )
+        # Merge configuration
+        ctx.default_map.update({k.replace("-", "_"): v for k, v in conf.items()})
```

Comment:
What do you think about the way this is done here @murilo-cunha? My main worry is that we're linking app configuration to `click`. What do you think about having a `settings.py` or simply a `dict` that we manage ourselves? The responsibility of `typer` would then be to properly populate this `dict`/`Settings` object and to then call the right command. And this command would then be another function, e.g., `command.meta`, which would be responsible simply for executing the core functionality given certain parameters. Decoupling the core logic and the configuration management+command arg parsing+etc would provide us with a bunch of advantages, here's a non-exhaustive list:
- We can later change the command parsing tool if we want to without having to touch the core logic
- We can directly call the core functions (e.g., `command.meta`) in our tests, without having to go through the CLI parsing
- We have the option of supporting other front-ends in the future (e.g., a UI or webserver, assuming that we ever want to go that route of course)
- It simplifies the code of the functions themselves (`cli.meta` and `command.meta` in this case): one just deals with parameter management and parsing basically, the other just deals with "business logic"
- We can call our config resolution in our tests without actually having to assess the impact it would have on the core logic (e.g., we can test that `overwrite = true` is set in a certain use case by just asserting that the resulting `dict` has `overwrite = true`, rather than actually running it and verifying that it has overwritten).

Reply:
If I understand you correctly, this is what I had started with, and I started `config.py` with that in mind. More specifically, I thought we could use [Pydantic models](https://pydantic-docs.helpmanual.io/usage/settings/) for this even (we can still add add the pydantic models later on to allow many file types). But I ran into some issues along the way.
So just to be clear, the my approach was:
1. Disregard the whole `is_eager` and `callback`
2. First thing that the `databooks.cli.{meta,fix}` does is to merge CLI arguments and config
3. Carry on with the rest of the program with the merged args
Also, the desired output is to have the parameter in following hierarchy (high priority to low):
1. CLI arguments (user passed arguments - ie: `--rm-outs`, etc.)
2. Config arguments (from `pyproject.toml`)
3. Default arguments

The issues that I had were:
1. Say we store the config in an object called `config` - we'd have to replace all instances of the `var` for `config.var`, which would make the code more polluted and require some hard coding (either specify different config objects for each command or specify a `config.command.var`
2. The first thing the command would do is find the config path (we need path that is passed in the CLI to find it), so the logic of injecting the config is with the logic of the command itself
3. **Most importantly,** if we do this in the command logic, then we cannot distinguish between what was defaults and what the user passed. So the config would always be overwritten by the default args (and break the order of hierarchy from before)

I tried different things and this was the cleanest solution. This is also the way that [`black` works](https://github.com/psf/black/blob/main/src/black/__init__.py#L110) (not that we have to do it the same, but when I was stuck and was comparing my ways out this seemed like the cleanest solution).

Final thoughts:
- This way I think we have the logic from the config almost entirely separate from the command (if you ignore the callbacks you'd be tricked into thinking that there is no config options)
- We can make the config logic very generic and we can reuse it
- This feels like using Click and it's features as it was intended. So we avoid reinventing the wheel.

Not sure I was missing something and if there is a more elegant way I'd be happy to refactor this!
Wdyt?

Reply:
Ah as for testing, I agree that maybe we should look into a way of testing the conf parsing more clearly. As I feel like this feature was made for this `ctx.default_map`, I'd imagine there is a clean way to test this.
I'll look more into it as well and add some tests if I find anything 😊

Reply:
Ok, I can see what you're saying, it makes sense. Let's see for the testing indeed, if there is nothing like this out of the box for typer then I guess it could still make sense to separate the business logic and argument parsing overall, then it would be something like:
```
@app.command
def meta(the_typer_args_here...):
    cmd.meta(the_business_args...) # don't include stuff like `--config` here, because it doesn't matter for the meta logic
```
and then you can monkeypatch `cmd.meta` to see what business args the `meta` typer command came up with and sent to the `meta` command. And similarly you can test `cmd.meta` without having to essentially pass CLI options. A disadvantage is that there'll be a lot of overlap with the `cli.meta` and `cmd.meta` function arguments, so a lot of copy-pasting I can imagine... But let's see then how `typer` recommends to do it indeed. I'll resolve this thread for now as it's either way not super relevant for the PR